iocore/net/QUICNetVConnection.cc

```diff
@@ -1594,6 +1594,21 @@ QUICNetVConnection::_packetize_frames(QUICEncryptionLevel level, uint64_t max_pa
     }
   }
 
+  // one ping per packet if needed.
+  if (!ack_eliciting && this->_pinger->will_generate_frame(level, timestamp)) {
+    frame = this->_pinger->generate_frame(frame_instance_buffer, level, this->_remote_flow_controller->credit(), max_frame_size,
```

Comment:
PING frame is different from others. It only have one ping per packets. And we don't need this ping frame when packet is `ack_eliciting`. So the ping count now change to how many `ack_eliciting` packets should be sent.

Not sure here. Is any better way to solve the case?

Reply:
I may be missing something, but if you just want to send one ping per packet, you can control it in `will_generate_frame()` by checking `timestamp`. You can return `false` if the last timestamp was the same as the current one. Actually I introduced `timestamp` for this, and some other frame generators use this logic to prevent sending multiple frames on one packet.

Reply:
You mean this timestamp
https://github.com/apache/trafficserver/blob/0afaec6934c4b2039a32b2817acded4bc1a4875d/iocore/net/QUICNetVConnection.cc#L1512-L1521

Actually it only returns a static variable and will be update when ats returns to eventsystem. So it is always the same value. That was design to decrease the some irqs (Not sure, but it seems the high rate time need to hardware support.if so it will involve the hard irq and soft irq). 

Of course you can use the `get_hrtime_updated` instead in this case.

Reply:
Oh, I knew it's not a real timestamp but didn't know the value is updated when it returns to the eventsystem. Well, then we need to do something, not only for this one but also other generators. We should avoid `get_hrtime_updated` because it's expensive.

Actually just a random value works. No generators need a timestamp. I just thought a timestamp might be useful for something.

Reply:
How about use tow generator queue. The random value might be expensive either.

Reply:
I meant meaningless values by random. A sequential number is fine as well. I think we can have `static uint8_t seq_num` in `_packetize_frames` and keep incrementing it.

What's two generator queue?

Reply:
Dispatch the generators into two queue. One repeatable generator, one unrepeatable. So we first walk through the repeatable queue then walk through unrepeatable queue. 

I'm ok with the seq_num but not in static (might be a race). It might be better to put it into `class QUICNetVConnection`

Reply:
Two queues can be problematic. Because the first queue would be like a priority queue, and the second queue might have less chance to send frames due to a limit of packet size. In current one queue design, we can prioritize frames (or generators) by importance but not repeatability.

A static variable can cause a race but I don't think it will be a real problem. The possibility that an instance read the same value in a row is really low. It happens only if the variable is incremented 256 times by other instances before the next read. Use of `uint64_t` make it more reliable. Having one uint64_t is more memory efficient than having uint8_t per connection.

Reply:
>The possibility that an instance read the same value in a row is really low.

I'm not sure that is correct way. IIUC, Non of the C/C++ operations is thread safe . For architecture, the cpu can not do anything directly in memory. It need to load the memory value and write it to register. Then increase value and write it back . Before write it back there are some cache(l1, l2, l3) between cpu and memory. So if you simply do ++ and update the memory. The other cpu might update an old value to memory and  override the new one. So next read might be an old value.

Reply:
Done

Reply:
I still think it's ok even if some cpu grab an old value.

Let's say currently `1` is in a variable, and two threads unfortunately grabbed the same value (`1`). Then they both increment it and `2` will be stored (twice). In general, it will be a problem because people usually expect the value is `3`. But here, we only expect that the value is not `1`.

Reply:
See

```
 cpu 1                    cpu2
---------------------------------------

 1. packetize                

 2. load seq                 

 3. increase                  

 4. writeback           packetize
 
 5. increase            load seq

 6. writeback           increase

 7. invalidate          writeback

 8. packetize           invalidate

 9. load seq (old value)

---------------------------------------


invalidate means refresh cache
writeback means write to l1cache
```

Sometimes kernel are not invalidate lcache immediately. So the var is always cached in lcache. In this case, cpu 2 read old value and do increasement then write it down to lcache, then refresh cache. So the data is 1 behind cpu 1. And the cpu 1 will read 1 behind the expect value.

Reply:
Can you put values beside each steps?

Why cpu 1 does increase twice?

If the initial value on step 1 is `1`, are you saying a read value can be `1` again on step 9?

Reply:
https://en.wikipedia.org/wiki/CPU_cache

>Alternatively, in a write-back or copy-back cache, writes are not immediately mirrored to the main memory, and the cache instead tracks which locations have been written over, marking them as dirty. The data in these locations is written back to the main memory only when that data is evicted from the cache. For this reason, a read miss in a write-back cache may sometimes require two memory accesses to service: one to first write the dirty location to main memory, and then another to read the new location from memory. 

This is from wiki, because we have no idea when and where kernel flush the cache to main memory.

Reply:
```

 cpu 1           cpu1 l1/memory   cpu2
------------------------------------------------------

 1. packetize         

 2. load seq 0    0/0      

 3. increase 1    1/0          

 4. writeback 1   1/0         packetize
 
 5. increase 2    2/0         load seq 0

 6. writeback 2   2/0         increase 1

 7. flush 2       2/2         writeback 1

 7. invalidate     /1         flush

 8. packetize                 invalidate

 9. load seq 1

---------------------------------------

```
I made a mistake the invalidate and flush are totally different.